modules/atomic-widgets/props-handler/props-handler.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+namespace Elementor\Modules\AtomicWidgets\PropsHandler;
+
+use Elementor\Modules\AtomicWidgets\Schema\Atomic_Prop;
+use Exception;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Props_Handler {
+	/**
+	 * Each transformer can return a value that is also a transformable value,
+	 * which means that it can be transformed again by another transformer.
+	 * This constant defines the maximum depth of transformations to avoid infinite loops.
+	 */
+	const TRANSFORM_DEPTH_LIMIT = 3;
+
+	const CONTEXT_SETTINGS = 'settings';
+	const CONTEXT_STYLE = 'style';
+
+	/**
+	 * @var array<Props_Handler>
+	 */
+	private static array $instances = [];
+
+	private Transformers_Registry $transformers;
+
+	public function __construct( Transformers_Registry $transformers ) {
+		$this->transformers = $transformers;
+	}
+
+	public static function instance( string $context, bool $fresh = false ): self {
+		if ( ! isset( self::$instances[ $context ] ) || $fresh ) {
+			$registry = new Transformers_Registry();
+
+			do_action( "elementor/atomic-widgets/{$context}/transformers", $registry );
+
+			self::$instances[ $context ] = new self( $registry );
+		}
+
+		return self::$instances[ $context ];
+	}
+
+	public static function for_style() {
+		return self::instance( self::CONTEXT_STYLE );
+	}
+
+	public static function for_settings() {
+		return self::instance( self::CONTEXT_SETTINGS );
+	}
+
+	public function handle( array $props, array $schema ): array {
+		$result = [];
+
+		foreach ( $schema as $prop_name => $prop_type ) {
+			$result[ $prop_name ] = $prop_type instanceof Atomic_Prop
+				? $this->transform( $props[ $prop_name ] ?? $prop_type->get_default(), $prop_type )
+				: null;
+		}
+
+		return $result;
+	}
+
+	private function transform( $value, Atomic_Prop $prop_type, int $depth = 0 ) {
+		if ( ! $value || ! $this->is_transformable( $value ) ) {
+			return $value;
+		}
+
+		if ( $depth >= self::TRANSFORM_DEPTH_LIMIT ) {
+			return null;
+		}
+
+		$transformer = $this->transformers->get( $value['$$type'] );
+
+		if ( ! ( $transformer instanceof Transformer_Base ) ) {
+			return null;
+		}
+
+		try {
+			$transformed_value = $transformer
+				->set_props_handler( $this )
+				->transform( $value['value'] ); // TODO: Here the prop_type should pass the children prop_types schema so it can be used in the transformer.
```

Comment:
you're reusing the same instance of the transformer for all of the transformations
I think it's kinda weird to do that and change the props handler each time

1. maybe `$transformers->get( ... )` should return a new instance?
2. can we somehow have a race condition where 2 processes will use the same transformer and one of them will override the props handler of the other one? (I know it's PHP and everything, just wondering)